### PR TITLE
feat(ena_deposition): Make deposition a package instead of multiple snakemake rules

### PR DESCRIPTION
**Pull request by anna-parker**
resolves https://github.com/loculus-project/loculus/issues/2878

preview URL: https://refactor-ena-deposition.loculus.org/

### Problem
If one snakemake rule fails the other rules will keep running indefinitely. We should stop other rules if one fails.

### Summary

- Create a new package called ena_deposition. One config is generated at startup, this uses defaults from the config/default.yaml and then overwrites additional config arguments from the passed config/config.yaml (as the snakefile did before). This reduces repetition as instead of one config for each rule, also there is now only one config and only one cli entry point.
- The `create_project`,`create_sample`, `create_assembly`, `trigger_submission` and `upload_external_metadata` functions are run in parallel threads (instead of before being run in parallel snakemake rules) and use stop_event to stop all threads if one fails. 
Tested that this works as intended on a preview by causing an Exception to be raised in the `trigger_submission` function thread - this lead to all other threads being stopped and the pod stopping and being restarted by kubernetes: 
```
13:33:20    DEBUG (  create_assembly.py: 455) - Checking state in ENA
Task generated an exception: Testing threads stop!!!
upload_external_metadata stopped due to exception in another task
create_sample stopped due to exception in another task
create_assembly stopped due to exception in another task
create_project stopped due to exception in another task
```

Additional changes:
- snakemake was removed as a dependency, this lead to `xmltodict` working differently than before. I needed to create a new class called `XmlNone` for the case when a field is intentionally left None in the dictionary, e.g. `SEQUENCING_PROJECT` in the xml below:
```xml
<SUBMISSION_PROJECT>
    <SEQUENCING_PROJECT/>
    <ORGANISM>
        <TAXON_ID>Test taxon</TAXON_ID>
        <SCIENTIFIC_NAME>Test scientific name</SCIENTIFIC_NAME>
    </ORGANISM>
</SUBMISSION_PROJECT>
```
- As I now specify the log level in __main__ I do not create a custom logger object for each script but use `logging` in all files - leading to `logger` being changed to `logging` in multiple files. 
- I created a `secure_ena_connection` function to check if connections to ENA are done correctly instead of doing this at the start of the snakefile.

### PR Checklist

- [x] All necessary documentation has been adapted.
- [x] Add trigger_submission_to_ena_from_file option
- [x] Test if one thread raises an error the other will stop
- [x] Test submission still works: tested full submission cycle including upload to Loculus

### CODE REVIEW

File: ena-submission/src/ena_deposition/config.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    if submit_to_ena_prod and (config.backend_url not in config.allowed_submission_hosts):
+        logging.warn("WARNING: backend_url not in allowed_hosts")
+        submit_to_ena_prod = False
+    submit_to_ena_dev = not submit_to_ena_prod
+
+    if submit_to_ena_dev:
+        config.test = True
+        logging.info("Submitting to ENA dev environment")
+        config.ena_submission_url = "https://wwwdev.ebi.ac.uk/ena/submit/drop-box/submit"
+        config.github_url = "https://raw.githubusercontent.com/pathoplexus/ena-submission/loculus_test/test/approved_ena_submission_list.json"
```
Comment (anna-parker) [inline ena-submission/src/ena_deposition/config.py:53]:
```suggestion
        config.github_url = "https://raw.githubusercontent.com/pathoplexus/ena-submission/main/test/approved_ena_submission_list.json"
```
revert before merging

File: .github/workflows/ena-submission-tests.yaml [hunk 2]
```diff
@@ -32,6 +32,7 @@ jobs:
       - name: Run tests
         run: |
             micromamba activate loculus-ena-submission
```
Comment (corneliusroemer) [inline .github/workflows/ena-submission-tests.yaml:34]:
```suggestion
```

File: ena-submission/README.md [hunk 3]
```diff
@@ -1,10 +1,10 @@
 # ENA Submission
 
-## Snakemake Rules
+## Cronjob
 
 ### get_ena_submission_list
 
-This rule runs daily in a cron job, it calls the loculus backend (`get-released-data`), obtains a new list of sequences that are ready for submission to ENA and sends this list as a compressed json file to our slack channel. Sequences are ready for submission IF:
+This script runs daily in a cron job, it calls the loculus backend (`get-released-data`), obtains a new list of sequences that are ready for submission to ENA and sends this list as a compressed json file to our slack channel. Sequences are ready for submission IF:
```
Comment (corneliusroemer) [inline ena-submission/README.md:7]:
```suggestion
This script runs once daily as a kubernetes cronjob. It calls the Loculus backend (`/get-released-data`), computes a new list of sequences that are ready for submission to ENA and sends this list as a compressed json file to our slack channel. Sequences are ready for submission IFF:
```

File: ena-submission/README.md [hunk 4]
```diff
@@ -13,9 +13,9 @@ This rule runs daily in a cron job, it calls the loculus backend (`get-released-
   - data is not in the `ena-submission.submission_table`
   - as an extra check we discard all sequences with `ena-specific-metadata` fields
 
-### all
+## Threads
 
-This rule runs in the ena-submission pod, it runs the following rules in parallel:
+The ena_deposition package, runs the following subfunctions in parallel:
```
Comment (corneliusroemer) [inline ena-submission/README.md:18]:
```suggestion
The ena_deposition package, runs the following functions in parallel (via threads):
```

File: ena-submission/src/ena_deposition/call_loculus.py [hunk 5]
```diff
@@ -182,100 +163,3 @@ def fetch_released_entries(config: Config, organism: str) -> dict[str, Any]:
...
-    log_level,
-    config_file,
-    output_file,
-    remove_if_has_metadata=False,
-):
-    """
-    Call Loculus.
-    """
-    logger.setLevel(log_level)
-    logging.getLogger("requests").setLevel(logging.WARNING)
```
Comment (corneliusroemer) [inline ena-submission/src/ena_deposition/call_loculus.py:237]: This is where we used to silence requests

**Review by corneliusroemer — APPROVED** [review]
Looks great, nice DRYification!

**Pull request merged**